채널에 참여인원이 몇명인지 확인할 수 있다.

happydhKim opened the pull request:
# 채널에 참여인원이 몇명인지 확인할 수 있다.

## 해당 이슈 📎

- [Epic#30] 채널에 참여인원이 몇명인지 확인할 수 있다.(#81)

## 변경 사항 🛠

- 채널 내 사용자를 카운트 할 수 있는 로직이 추가되었습니다.

## 테스트 ✨

> 없음

## 리뷰어 참고 사항 🙋‍♀️

- apollo-server의 app.js에서 분리를 어떤식으로 할지 고민을 해보았으나, 괜찮은 방법을 찾지 못했습니다.
- 어떤식으로 분리하면 좋을지 좋은 방안이 있으시다면, 꼭 리뷰로 부탁드립니다.

happydhKim committed:
[ FEATURE ] 채널 정보에 몇명이 있는지 나타내는 컴포넌트 구현 (#81)

Why
- 채널 정보에 몇명이 접속해 있는지 나타내는 컴포넌트가 필요하다.

What
- 채널 타이틀 아래 몇명이 있는지 나타낼 수 있는 People 관련 컴포넌트 구현.
- 사람 모형의 이모지인 people.svg를 public 폴더에 추가.

happydhKim committed:
[ FEATURE ] 채널 내 유저 카운팅을 위한 스키마 추가 (#81)

Why
- 채널 내 유저를 카운팅 하기 위해 새로운 channel 칼럼이 필요하다.

What
- 유저를 카운팅 하기 위해 새로운 mutation을 생성.
- 유저를 카운팅 하기 위해 channel에 새로운 컬럼 usercount를 생성.
- 유저를 카운팅 하기 위해 새로운 hook을 생성.

happydhKim committed:
[ FEATURE ] 사용자가 채널을 나가는 이벤트 추가 (#81)

Why
- 사용자가 채널을 나가는 순간을 캐치해야지 count가 가능하다.

What
- beforeunload 이벤트를 사용할 수 있도록 공통 component에 파일 생성.
- beforeunload를 사용할 수 있도록 코드 구현.

happydhKim committed:
[ FEATURE ] 사용자가 채널에 들어오고 나갈 때, count되도록 구현 (#81)

Why
- 사용자가 채널에 나가고 들어올 때, 총 인원을 갱신해줘야 한다.

What
- channel 스키마에 userCount 수정 ( number -> Int )
- 인원갱신 mutation을 사용하기 위한 hook 생성. ( useSetUserCount )
- 들어오고 나가는 순간을 hook으로 구별해 갱신함.
- 전역 페이지에 channel을 생성. (PLUS, MINUS)

happydhKim committed:
[ REFACTOR ] 기존 채널 내 사용자를 카운팅 하는 방식을 배열로 변경 (#81)

Why
- 채널 내 사용자를 숫자로 카운팅 하면, 처리가 불가능하다.

What
- 채널 내 사용자의 아이디 값을 배열로 저장해서 누가 나갔는지 확인할 수 있도록 구현.
- 채널 스키마를 수정.

happydhKim committed:
[ REFACTOR ] 유저 카운팅 서버사이드 리펙토링 (#81)

Why
- 스키마 오류로 채널 초기에 데이터를 불러오지 못해 재설계가 필요함.

What
- 스키마 재설계를 했지만, 급한 다른 작업이 생겨 작업 일시 중단.

happydhKim committed:
[ CHORE ] 채널 내 사용자 카운트 프론트 작업 임시 커밋 (#81)

Why
- 임시 저장을 해둬야 한다.

What
- 우선시 되는 채널 작업을 위해 중단을 하고 임시 저장을 한다.

happydhKim committed:
[ MERGE ] Develop 상태 Feature 브랜치와 동기화

* develop -> feature/issue81

Why
- 중간에 보류되었던 작업을 재작업하기 위해서

What
- Develop 브랜치 feature/issue81 브랜치에 머지

happydhKim committed:
[ FEATURE ] 채널 내 인원 카운트 useBeforeunload 이벤트 적용 (#81)

Why
- 채널 내 인원 카운트를 ws으로 처리하기 곤란해 브라우저 이벤트를 사용해야 한다.
- 하지만 간혹 이벤트가 작동이 안되는 경우가 있어 다른 ws방안을 고려 해야 한다.

What
- useBoforeunload 이벤트를 연동.
- 테스트를 진행 했으나 불안정.

happydhKim committed:
[ MERGE ] 최신 버전으로 업데이트 머지 (#81)

Why
- 새롭게 배포된 버전으로 업데이트를 해야 한다.

What
- 최신 버전으로 업데이트.

happydhKim committed:
[ MERGE ] 최신버전으로 업데이트 (#81)

Why
- 최신으로 업데이트 된 파일과 합쳐서 작업을 해야 한다.

What
- 최신 버전의 develop 브랜치와 병합.

happydhKim committed:
[ MERGE ] 최신 버전으로 업데이트 (#81)

Why
- 최신버전의 코드로 업데이트 해야 한다.

What
- 최신버전 develop 코드로 업데이트.

happydhKim committed:
[ MERGE ] 최신버전으로 동기화 (#81)

Why
- 최신 버전으로 동기화 해야 한다.

What
- 최신 버전의 devleop으로 동기화.

happydhKim committed:
[ FEATURE ] listener typedefs 재정의 (#81)

Why
- listener list typedefs를 재정의 해야 한다.

what
- listener list typedfs 재정의.

happydhKim committed:
[ MERGE ] 최신 버전으로 업데이트 (81)

happydhKim committed:
[ FEATURE ] 채널 내 사용자 카운트 DB연동 (#81)

Why
- 데이터베이스를 이용해 유저 수를 관리해야 한다.
- app.js에서 ApolloServer를 이용해 구현해야 한다.

What
- subscriptions 을 이용해서 onConnect 와 disConnect를 이용해 구현한다.
- 클라이언트에서 Entrance에 listenerList를 추가했다.

happydhKim committed:
[ FEATURE ] 채널 내 사용자 카운트 동기화 적용 (#81)

Why
- 채널 내 사용자 카운트를 동기화 시켜야 한다.

What
- useListenerListChanged 훅을 사용해서 동기화를 적용.
- apollo-server app.js 에서 pubsub을 통해서 통신함.
- app.js 리팩토링을 해야함. 데이터베이스를 app.js에서 직접 사용함.

happydhKim committed:
[ MERGE ] 최신 버전의 코드로 동기화 (#81)

Why
- 최신버전의 코드로 동기화를 시켜야 한다.

What
- 최신버전의 develop 코드로 동기화.

happydhKim committed:
[ REFACTOR ] 기존 app.js에서 중복되는 부분 함수로 빼냄 (#81)

Why
- 중복되는 부분을 함수로 빼내야 한다.

What
- 중복되는 부분을 모두 함수로 뺴냄.

always-awake commented on web/src/components/channel/Slide/SlideInfo/SlideInfo.jsx:
```diff
@@ -10,13 +11,22 @@ const SlideInfo = () => {
+            src={People}
+            alt="people-logo"
+          />
+          120명 참여중
```
인원 수가 하드코딩 되어 있는건가요??

always-awake commented on web/src/hooks/channel/useSetUserCount.js:
```diff
@@ -2,19 +2,20 @@ import gql from 'graphql-tag';
+    ? result.data.userCount
     : null;
-
+  // const [setCurrentSlide] = useMutation(SET_CURRENT_SLIDE);
```
주석 제거해주세요

always-awake commented on web/src/hooks/channel/useSetUserCount.js:
```diff
@@ -2,19 +2,20 @@ import gql from 'graphql-tag';
-  mutation useSetUserCount($channelId: String! ,$userCount: Number!) {
-    addUser(channelId: $channelId, userCount: $userCount) { 
-      countUser
+  mutation UseSetUserCount($channelId: String! ,$userCount: Int!) {
```
mutation 사용하는 컨벤션 지켜주시면 좋을거 같아요~
다른 파일 보시면 잘 나와있을거에여!

always-awake commented on server-api/src/apollo/typeDefs/channels.js:
```diff
@@ -9,7 +9,7 @@ type Channel {
   channelStatus: String
   currentSlide: Int!
-  userCount: Int!
+  userCount: [String]
```
이후에 userCount = userCount.length 식으로 사용하시는데 useCount 대신에 listeners, listenerList 처럼 변경해주시는게 어떨까요?

happydhKim committed:
[ REFACTOR ] Apollo-server app.js 의 socket로직 리팩토링 (#81)

Why
- app.js가 방대해져서 분리를 해야 한다.

What
- socket 로직을 분리해서 webSocket폴더로 이동.
- 유저를 카운트 하는 부분을 빼내서 관리.
- 채널 스피커의 입장을 확인하는 로직을 뺴내서 관리.

happydhKim committed:
[ MERGE ] 코드를 최신으로 업데이트 (#81)

Why
- 코드를 최신버전으로 동기화 해야 한다.

What
- 최신버전의 develop 코드로 업데이트.

always-awake reviewed (approved):
코드가 정말 깔끔해졌습니다 👍 수고 많으셨어요!
이전에 했던 리뷰에 해당하는 코드는 이미 다 수정하신것 같네요. 수고 많으셨습니다

happydhKim committed:
[ FIX ] 가독성을 위한 수정 (#81)

Why
- 가독성을 높이기 위해 줄을 띄워줄 필요가 있다.

What
- 가독성을 높이는 빈줄 추가.
